File: sdk/python3/nmidsdk/worker/response.py
import struct
import nmidsdk.worker.utils as utils
import nmidsdk.model.const as const

class Response:
    def __init__(self):
        self.data_type = 0
        self.data = bytearray()
        self.data_len = 0
        self.handle = ""
        self.handle_len = 0
        self.params_type = 0
        self.params_handle_type = 0
        self.params_len = 0
        self.params = bytearray()
        self.params_map = {}
        self.job_id = ""
        self.job_id_len = 0
        self.ret = bytearray()
        self.ret_len = 0
        self.agent = None

    def decode_pack(self, data: bytearray):
        res_len = len(data)
        if res_len < const.MIN_DATA_SIZE:
            return None, res_len, ConnectionError("Invalid data length")

        cl, = struct.unpack('>I', data[8:const.MIN_DATA_SIZE])
        if res_len < const.MIN_DATA_SIZE + cl:
            return None, res_len, ConnectionError("Invalid data content length")

        content = data[const.MIN_DATA_SIZE:const.MIN_DATA_SIZE+cl]
        if len(content) != cl:
            return None, res_len, ConnectionError("Invalid data content length")

        resp = Response()
        resp.data_type, = struct.unpack('>I', data[4:8])
        resp.data_len = cl
        resp.data = content

        if resp.data_type == const.PDT_S_GET_DATA:
            start = const.MIN_DATA_SIZE
            resp.params_type, = struct.unpack('>I', data[start:start+const.UINT32_SIZE])
            start += const.UINT32_SIZE
            resp.params_handle_type, = struct.unpack('>I', data[start:start+const.UINT32_SIZE])
            start += const.UINT32_SIZE
            resp.handle_len, = struct.unpack('>I', data[start:start+const.UINT32_SIZE])
            start += const.UINT32_SIZE
            resp.params_len, = struct.unpack('>I', data[start:start+const.UINT32_SIZE])
            start += const.UINT32_SIZE
            resp.job_id_len, = struct.unpack('>I', data[start:start+const.UINT32_SIZE])
            
            start += const.UINT32_SIZE
            resp.handle = data[start:start+resp.handle_len].decode('utf-8')
            start += resp.handle_len
            resp.parse_params(data[start:start+resp.params_len])
            start += resp.params_len
            resp.job_id = data[start:start+resp.job_id_len].decode('utf-8')

        return resp, res_len, None
    
    def parse_params(self, params):
        self.params = params
        if self.params_type == const.PARAMS_TYPE_MSGPACK:
            self.params_map = utils.msgpack_params_map(params)
        if self.params_type == const.PARAMS_TYPE_JSON:
            self.params_map = utils.json_params_map(params)

    def get_response(self):
        return self
    
    def should_bind(self, obj: dict):
        obj.update(self.params_map)
        
    def get_params(self):
        if self.params_len == 0:
            return None
        
        return self.params
    
    def get_params_map(self):
        if self.params_len == 0:
            return None
        
        return self.params_map

File: sdk/node/lib/worker/response.js
'use strict';

// Worker response decoder.
// Mirrors sdk/python3/nmidsdk/worker/response.py and pkg/worker/response.go.
//
// For PDT_S_GET_DATA the body (read from the packet at MIN_DATA_SIZE) is:
//   ParamsType(4) | ParamsHandleType(4) | HandleLen(4) | ParamsLen(4) |
//   JobIdLen(4) | Handle | Params | JobId
// Params are parsed (msgpack or json) into paramsMap.

const C = require('../const');
const { msgpackParamsMap, jsonParamsMap } = require('./utils');

class Response {
  constructor() {
    this.dataType = 0;
    this.data = Buffer.alloc(0);
    this.dataLen = 0;

    this.handle = '';
    this.handleLen = 0;
    this.paramsType = 0;
    this.paramsHandleType = 0;
    this.paramsLen = 0;
    this.params = Buffer.alloc(0);
    this.paramsMap = null;
    this.jobId = '';
    this.jobIdLen = 0;

    this.ret = Buffer.alloc(0);
    this.retLen = 0;

    this.agent = null; // back-reference to the Agent that received this job
  }

  getResponse() {
    return this;
  }

  getParams() {
    if (this.paramsLen === 0) return null;
    return this.params;
  }

  getParamsMap() {
    if (this.paramsLen === 0) return null;
    return this.paramsMap;
  }

  // Merge paramsMap into the given object, mirroring python should_bind.
  shouldBind(obj) {
    if (this.paramsMap) Object.assign(obj, this.paramsMap);
  }

  parseParams(params) {
    this.params = params;
    if (this.paramsType === C.PARAMS_TYPE_MSGPACK) {
      this.paramsMap = msgpackParamsMap(params);
    } else if (this.paramsType === C.PARAMS_TYPE_JSON) {
      this.paramsMap = jsonParamsMap(params);
    }
  }
}

// Decode a single complete packet (>= MIN_DATA_SIZE + dataLen) into a
// Response. Returns { resp, err }.
function decodePack(data) {
  const resLen = data.length;
  if (resLen < C.MIN_DATA_SIZE) {
    return { resp: null, err: new Error('Invalid data: too short') };
  }

  const cl = data.readUInt32BE(8); // content length
  if (resLen < C.MIN_DATA_SIZE + cl) {
    return { resp: null, err: new Error('Invalid data: incomplete body') };
  }

  const content = data.subarray(C.MIN_DATA_SIZE, C.MIN_DATA_SIZE + cl);
  if (content.length !== cl) {
    return { resp: null, err: new Error('Invalid data: content length mismatch') };
  }

  const resp = new Response();
  resp.dataType = data.readUInt32BE(4);
  resp.dataLen = cl;
  resp.data = content;

  if (resp.dataType === C.PDT_S_GET_DATA) {
    let start = C.MIN_DATA_SIZE;
    resp.paramsType = data.readUInt32BE(start);
    start += C.UINT32_SIZE;
    resp.paramsHandleType = data.readUInt32BE(start);
    start += C.UINT32_SIZE;
    resp.handleLen = data.readUInt32BE(start);
    start += C.UINT32_SIZE;
    resp.paramsLen = data.readUInt32BE(start);
    start += C.UINT32_SIZE;
    resp.jobIdLen = data.readUInt32BE(start);
    start += C.UINT32_SIZE;

    resp.handle = data.toString('utf8', start, start + resp.handleLen);
    start += resp.handleLen;

    resp.parseParams(data.subarray(start, start + resp.paramsLen));
    start += resp.paramsLen;

    resp.jobId = data.toString('utf8', start, start + resp.jobIdLen);
  }

  return { resp, err: null };
}

module.exports = { Response, decodePack };
